### DISCUSSION TITLE
my array doesn't rerender when its value changed

### DISCUSSION THREAD

**Discussion by deli-ght** [pmndrs/jotai#1664] [Q&A] [answered]
```
export const userAnswerAtom = atom<string[]>(Array.from(new Array(LENGTH_OF_WORD), () => ""));

export const updateUserAnswerAtom = atom<string[], string>(
  (get) => get(userAnswerAtom),
  (get, set, update) => {
    const updatedAnswer = get(userAnswerAtom);

    get(answerAtom)
      .split("")
      .forEach((alphabet: string, index: number) => {
        if (alphabet === update) {
          updatedAnswer[index] = update;
        }
      });
    set(userAnswerAtom, updatedAnswer);
  },
);

```
 
I set my atom like this then change the value using setUpdateUserAnswerAtom.
the value is actually changed when i print it on console, but the component using this atom doesn't rerendered so that I couldn't check updated value. it there any way to rerender that component when this atom changed?
i added another set function to check rerender and it works but only array atom doesn't rerender when it changed!

**Comment by dai-shi** [comment, marked as answer]
You can't mutate atom values. Update them immutably instead.

An easy solution:
```diff
-     const updatedAnswer = get(userAnswerAtom);
+     const updatedAnswer = [...get(userAnswerAtom)];
```

**Reply by deli-ght** [reply]
Thanks a lot! it works

**Reply by Johnrobmiller** [reply]
I don't think that reallocating an entire array in memory just to rerender the react component is the solution here. Maybe for small arrays that would be fine, but this is going to fall apart for anything bigger, causing memory fragmentations. This also doesn't work well when using Maps, as it is much more computationally expensive to recrate a map than it is to reallocate an array in memory

**Reply by dmaskasky** [reply]
Wow, that would be a really long answer word.

If reallocating an array is causing performance issues, there is likely a better way to design your data structures. For example, if the array was a tree of nested segments you could just spread the changed array segments.

Another approach is to use a symbol to coincide with the mutated array.
```ts
const symbolAtom = atom(Symbol())
const arrayBaseAtom = atom([])
const arrayAtom = atom(
  (get) => {
    get(symbolAtom)
    return get(arrayBaseAtom)
  },
  (get, set, update) => {
    const originalArray = get(arrayBaseAtom)
    if (!isSameArray(originalArray, update)) {
      set(symbolAtom, Symbol())
      Object.assign(originalArray, update)
    }
  }
)
```

**Reply by Johnrobmiller** [reply]
I may be wrong, but I don't think that data structures have anything to do with this. When your state needs to be in the form of an array, map, etc., then your state management library shouldn't stop you just because it can't handle them internally and requires the dev to use flags/symbols to track mutations

Regarding performance, if only the reference (i.e., the pointer) to an array changes, then the array contents do not have to remove themselves from the heap, causing heap fragmentations. However, if the array is recreated by doing `[ ...arrayContent ]`, then the bytes written to RAM have to be rewritten to somewhere else on the heap. Not only is this terrible performance in of itself, but it also causes the heap to become fragmented. It also accumulates memory garbage, which will cause weird glitches and lag as the garbage collection has to clean up the mess the developer made. 

I do not think it's good to recommend to people to do this.

Performance doesn't always matter depending on the project, but sometimes it does. You can't assume that people are not doing things on animation frames or working with large data on the front end. To be rather blunt, I think the willingness for javascript devs to recklessly reallocate arrays whenever they feel like it showcases why all the negative stereotypes exist about javascript devs to begin with. They often don't take these factors into consideration when writing code, which sometimes gets them in trouble depending on the project

**Reply by dmaskasky** [reply]
There's a few points I'd like to clarify
1. Shallow copies are cheap. [...array] copies only the reference slots, not the nested objects/strings. For an array of a few dozen or even thousands of items, this is microseconds of work.

2. Modern GC handles fragmentation. JavaScript engines use generational garbage collectors that move objects around; they do not leave the heap riddled with holes the way manual malloc/free in C might.

3. Comparing to Map: Creating new Map(oldMap) copies entry pointers, not the key/value objects themselves. It is still O(n), but in UI code the bottleneck is almost always rendering, not these copies.

For reference, the data structure I was referring to in my previous post is a chunked rope array
```ts
const CHUNK = 32;

type Leaf<T> = T[];
type Node<T> = Leaf<T> | Node<T>[];

function setAt<T>(
  root: Node<T>,
  index: number,
  value: T,
  depth = Math.ceil(Math.log2(rootSize) / Math.log2(CHUNK))
): Node<T> {
  if (depth === 0) {
    // leaf: copy the chunk, mutate the copy
    const leaf = [...(root as Leaf<T>)];
    leaf[index & (CHUNK - 1)] = value;
    return leaf;
  }

  const node = [...(root as Node<T>[])];
  const childIdx = (index >> (depth * Math.log2(CHUNK))) & (CHUNK - 1);
  node[childIdx] = setAt(node[childIdx], index, value, depth - 1);
  return node;
}
```

If an array is so large that a shallow copy is measurably slow, that state is probably better kept outside React/Jotai, and you’d pass derived summaries into atoms instead.

The choice of data structure is core to how libraries like Jotai detect changes and trigger re-renders: they compare the **reference identity** of the value you pass to `set`. If you mutate an array in place, its reference stays the same and React sees no change; if you use an immutable pattern - whether a simple spread (`[...arr]`) or a structurally-shared tree that only copies the nodes along the path you updated - you give Jotai a *new* reference and only the components that rely on the altered branch re-render.

Modern JavaScript engines compact the heap, shallow-copying even large arrays incurs negligible overhead, and persistent data structures (or Immer) automate minimal copying for you. The data structure directly affects update granularity, rendering cost, and garbage-collection behavior, so saying it “has nothing to do with this” overlooks the very mechanism React and Jotai use to decide whether to repaint the UI.

That said, I still provided a slick Symbol solution which addresses all of your concerns.

**Reply by Johnrobmiller** [reply]
I did some research now on the topic, and what you say is true except in the case where the array consists of primatives (i.e., numbers, single characters, or structs in other languages).

I'll admit that there are probably very few people who are needing to mutate bigger data structures (and also possibly on an animation frame), which is possibly the only scenario where this matters. And you are right that React doesn't seem to be designed for this niche audience. I looked into the issue of mutating bigger data structures in react, and what you say is right, React itself doesn't seem to suppor this very well, as there is no native way to mutate a data structure directly in a way that triggers a rerender (SolidJS has a solution to this, I found out)

I apologize if I came across as argumentative, and I do appreciate you giving your insight into all of this.
